### DISCUSSION TITLE
multi gpu function

### DISCUSSION THREAD

**Discussion by 54x1** [cdgriffith/FastFlix#558] [Ideas]
![image](https://github.com/cdgriffith/FastFlix/assets/26955286/abc0a8f6-9cb5-46b4-8b4c-af85d4129b1b)

in future can you list the gpus or have a multi gpu function?

**Comment by MarcoRavich** [comment]
Hi there, yesterday I've tested encoding performances - both with NVEnc and QSVEnc - on an a friend's [MSI Stealth GS77 - 12U](https://it.msi.com/Laptop/Stealth-GS77-12UX/Specification) laptop: when the process was done with one of the GPUs, the other remained unused - although it still consumed some power (neither goes totally idle) and of course produce heat - and vice versa.

Therefore it would be extremely interesting (and maybe not too difficult to implement) to allow user choose which GPU allocate for the decoding workload regardless of which has chosed for the encoding!

An interesting study about:
- [HEVC hardware vs software decoding: An objective energy consumption analysis and comparison](https://www.sciencedirect.com/science/article/abs/pii/S1383762121000199) paper claims _"The obtained experimental results revealed that, on the tested platforms, from the operating system level point of view, and for 1080p video resolution and lower, SW video decoding consumes on average less than 4 x more power as compared to HW video decoding. This gap is not as large as expected, according to the state of the art, thanks to SW parallel processing. **This is not true for high video data rate (bitrate, frame rate, and resolution) where the HW video decoding offers the best energy efficiency (e.g., a ratio of more than 10 x for video resolution). Thus, the ratio between the SW and HW HEVC decoding energy as a function of video resolution is not linear.**"_

About quality:
- According to [this post on KODI forum](https://forum.kodi.tv/showthread.php?tid=301258&pid=2481469#pid2481469) _"For quality purposes, assuming we are not factoring in any kind of 'post processing' effects, no, there will be no difference. Video compression codecs are designed to all decode the same result. If one decoder gave a different output than another, that would mean something was implemented incorrectly.
**If hardware decoding is possible, use hardware decoding and leave software decoding for anything that the hardware decoder doesn't support.** This is the most efficient way to go in terms of processor resources and electrical consumption."_

Last but not least there's also this old (but interesting anyway) [article @ Tom's Hardware](https://www.tomshardware.com/reviews/video-transcoding-amd-app-nvidia-cuda-intel-quicksync,2839.html).

Hope @cdgriffith will consider this.


EDIT
Some other maybe inspiring resources:
- https://www.reddit.com/r/ffmpeg/comments/10dejz2/will_hardware_decoding_affect_quality/
- https://superuser.com/questions/1712288/using-nvidia-nvdec-and-intel-qsv-encode
- @tumutech's [miniproject2](https://github.com/tumutech/miniproject2) library project;
- @bmegli's [hardware-video-decoder](https://github.com/bmegli/hardware-video-decoder) library project;
- @Proryanator's [encoder-benchmark ](https://github.com/Proryanator/encoder-benchmark) tool.

**Reply by MarcoRavich** [reply]
Linked to @Anime4000's [IFME](https://github.com/Anime4000/IFME/) "[discussion](https://github.com/Anime4000/IFME/discussions/245)" too...
